client/retryable_client.go

```diff
@@ -0,0 +1,279 @@
+package client
+
+import (
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/mysteriumnetwork/payments/bindings"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+)
+
+type blockchain interface {
+	GetAccountantFee(accountantAddress common.Address) (uint16, error)
+	IsRegisteredAsProvider(accountantAddress, registryAddress, addressToCheck common.Address) (bool, error)
+	GetProviderChannel(accountantAddress common.Address, addressToCheck common.Address) (ProviderChannel, error)
+	IsRegistered(registryAddress, addressToCheck common.Address) (bool, error)
+	SubscribeToPromiseSettledEvent(providerID, accountantID common.Address) (sink chan *bindings.AccountantImplementationPromiseSettled, cancel func(), err error)
+	GetMystBalance(mystSCAddress, address common.Address) (*big.Int, error)
+	SubscribeToConsumerBalanceEvent(channel, mystSCAddress common.Address) (chan *bindings.MystTokenTransfer, func(), error)
+	GetRegistrationFee(registryAddress common.Address) (*big.Int, error)
+	RegisterIdentity(rr RegistrationRequest) (*types.Transaction, error)
+	TransferMyst(req TransferRequest) (tx *types.Transaction, err error)
+	IsAccountantRegistered(registryAddress, acccountantID common.Address) (bool, error)
+	GetAccountantOperator(accountantID common.Address) (common.Address, error)
+	SettleAndRebalance(req SettleAndRebalanceRequest) (*types.Transaction, error)
+	GetChannel(acc common.Address, chID []byte) (ProviderChannel, error)
+}
+
+// BlockchainWithRetries takes in the plain blockchain implementation and exposes methods that will retry the underlying bc methods before giving up.
+// This is required as the ethereum client will occasionally spit a TLS error if running for prolonged periods of time.
+type BlockchainWithRetries struct {
+	delay      time.Duration
+	maxRetries int
+	bc         blockchain
+	stop       chan struct{}
+	once       sync.Once
+}
+
+// ErrStopped represents an error when a call is interrupted
+var ErrStopped = errors.New("call stopped")
+
+// NewBlockchainWithRetries returns a new instance of blockchain with retries
+func NewBlockchainWithRetries(bc blockchain, delay time.Duration, maxRetries int) *BlockchainWithRetries {
+	return &BlockchainWithRetries{
+		bc:         bc,
+		delay:      delay,
+		maxRetries: maxRetries,
+	}
+}
+
+func (bwr *BlockchainWithRetries) callWithRetry(f func() error) error {
+	for i := 0; i < bwr.maxRetries; i++ {
```

Comment:
For retry loops I suggest to start i from 1 which is more clear and you don't need extra math with `i + 1` in any place.

```go
for i := 1; i <= bwr.maxRetries; i++ {
    if i == bwr.maxRetries {
        return err
    }
}
```